pkg/util/soft_append_error_processor.go

```diff
@@ -0,0 +1,111 @@
+package util
+
+import (
+	"github.com/pkg/errors"
+	"github.com/prometheus/prometheus/model/histogram"
+	"github.com/prometheus/prometheus/storage"
+
+	"github.com/grafana/mimir/pkg/mimirpb"
+	"github.com/grafana/mimir/pkg/util/globalerror"
+)
+
+// SoftAppendErrorProcessor helps identify soft errors and run appropriate callbacks.
+// This also helps keep the soft error checks consistent between ingesters and block builders
+// by keeping it all in a single place.
+type SoftAppendErrorProcessor struct {
+	// CommonCallback is called irrespective of soft or hard error.
+	CommonCallback                   func()
+	ErrOutOfBounds                   func(ts int64, labels []mimirpb.LabelAdapter)
+	ErrOutOfOrderSample              func(ts int64, labels []mimirpb.LabelAdapter)
+	ErrTooOldSample                  func(ts int64, labels []mimirpb.LabelAdapter)
+	SampleTooFarInFuture             func(ts int64, labels []mimirpb.LabelAdapter)
+	ErrDuplicateSampleForTimestamp   func(ts int64, labels []mimirpb.LabelAdapter)
+	MaxSeriesPerUser                 func()
+	MaxSeriesPerMetric               func(labels []mimirpb.LabelAdapter)
+	ErrOOONativeHistogramsDisabled   func(ts int64, labels []mimirpb.LabelAdapter)
+	ErrHistogramCountMismatch        func(ts int64, labels []mimirpb.LabelAdapter)
+	ErrHistogramCountNotBigEnough    func(ts int64, labels []mimirpb.LabelAdapter)
+	ErrHistogramNegativeBucketCount  func(ts int64, labels []mimirpb.LabelAdapter)
+	ErrHistogramSpanNegativeOffset   func(ts int64, labels []mimirpb.LabelAdapter)
+	ErrHistogramSpansBucketsMismatch func(ts int64, labels []mimirpb.LabelAdapter)
+}
+
+// ProcessErr returns true if the err is a soft append error and calls appropriate callback function.
+// In case a soft error is encountered, we can continue ingesting other samples without aborting
+// the whole request. ProcessErr always calls the CommonCallback() if it exists.
+// err must be non-nil.
+func (e *SoftAppendErrorProcessor) ProcessErr(err error, ts int64, labels []mimirpb.LabelAdapter) bool {
+	if e.CommonCallback != nil {
+		e.CommonCallback()
+	}
+	switch {
+	case errors.Is(err, storage.ErrOutOfBounds):
+		if e.ErrOutOfBounds != nil {
```

Comment:
I think code would be more future-proof if we don't check for `nil` and we explicitly require that all callbacks are set in `SoftAppendErrorProcessor`. If an error doesn't have to be processed intentionally, we can just pass a no-op function.